File: Baseline_and_Metric/Clustering/Baselines/Python/CIMLR/run_cimlr.py
#!/usr/bin/env python3
"""
run_cimlr.py — CIMLR multi-omics clustering on MLOmics data

Python implementation of:
  CIMLR (Cancer Integration via Multikernel LeaRning)
  Rappoport & Shamir, NAR 2018 / Wang et al. Nature Methods 2017

Usage:
    python run_cimlr.py --dataset ACC --version Top \
        --data_path /path/to/Clustering_datasets --k 3 \
        --output_dir /path/to/output

Output:
    {output_dir}/results_{dataset}_{version}_CIMLR.csv
"""

import argparse, os, sys

def main():
    parser = argparse.ArgumentParser(description='CIMLR multi-omics clustering')
    parser.add_argument('--dataset',    required=True, help='Dataset name (e.g. ACC)')
    parser.add_argument('--version',    required=True, help='Data version: Top, Original, Aligned')
    parser.add_argument('--data_path',  required=True, help='Root path to Clustering_datasets/')
    parser.add_argument('--k',          type=int, default=2, help='Number of clusters (default: 2)')
    parser.add_argument('--n_neighbors',type=int, default=10, help='CIMLR k-neighbours (default: 10)')
    parser.add_argument('--max_iter',   type=int, default=30, help='Optimisation iterations (default: 30)')
    parser.add_argument('--output_dir', default='.', help='Output directory (default: .)')
    args = parser.parse_args()

    import numpy as np
    import pandas as pd

    # Locate this script's directory to import cimlr.py
    script_dir = os.path.dirname(os.path.abspath(__file__))
    if script_dir not in sys.path:
        sys.path.insert(0, script_dir)
    from cimlr import CIMLR

    # ── File suffix ──────────────────────────────────────────────────────────
    suffix_map = {'Original': '', 'Top': '_top', 'Aligned': '_aligned'}
    if args.version not in suffix_map:
        raise ValueError(f"Unknown version '{args.version}'")
    suffix = suffix_map[args.version]

    # ── Load omics data ──────────────────────────────────────────────────────
    base_dir = os.path.join(args.data_path, args.dataset, args.version)
    omics_types = ['mRNA', 'miRNA', 'CNV', 'Methy']
    omics_list  = []
    sample_ids  = None

    for ot in omics_types:
        fpath = os.path.join(base_dir, f'{args.dataset}_{ot}{suffix}.csv')
        if not os.path.exists(fpath):
            # Try without dataset prefix (clustering datasets use cancer code)
            cancer = args.dataset  # e.g. ACC
            fpath  = os.path.join(base_dir, f'{cancer}_{ot}{suffix}.csv')
        if not os.path.exists(fpath):
            print(f'  Skipping {ot}: file not found at {fpath}')
            continue
        df = pd.read_csv(fpath, index_col=0)
        # features × samples → samples × features
        X = df.T.values.astype('float32')
        X = np.nan_to_num(X, nan=0.0)
        if sample_ids is None:
            sample_ids = list(df.columns)
        print(f'  Loaded {ot}: {X.shape}')
        omics_list.append(X)

    if not omics_list:
        raise RuntimeError('No omics files found — check data_path and dataset name')

    print(f'Running CIMLR: {len(omics_list)} omics, {omics_list[0].shape[0]} samples, '
          f'k={args.k}, n_neighbors={args.n_neighbors}, max_iter={args.max_iter}')

    # ── Run CIMLR ────────────────────────────────────────────────────────────
    model = CIMLR(n_clusters=args.k, n_neighbors=args.n_neighbors,
                  max_iter=args.max_iter)
    labels, S, F = model.fit_predict(omics_list)

    # ── Save results ─────────────────────────────────────────────────────────
    os.makedirs(args.output_dir, exist_ok=True)
    out_file = os.path.join(args.output_dir,
                            f'results_{args.dataset}_{args.version}_CIMLR.csv')
    result_df = pd.DataFrame({'sample': sample_ids, 'cluster': labels})
    result_df.to_csv(out_file, index=False)
    print(f'Results saved to: {out_file}')
    print(f'Cluster distribution: {dict(zip(*np.unique(labels, return_counts=True)))}')


if __name__ == '__main__':
    main()

File: Baseline_and_Metric/Clustering/Baselines/R/CIMLR/run_CIMLR.R
#!/usr/bin/env Rscript
# run_CIMLR.R — Wrapper for CIMLR multi-omics clustering on MLOmics data
#
# CIMLR R package must be installed:
#   devtools::install_github("danro9685/CIMLR", ref = "R")
#
# Usage:
#   Rscript run_CIMLR.R --dataset ACC --version Top --data_path /path/to/Clustering_datasets --k 2
#
# Outputs:
#   results_{dataset}_{version}_CIMLR.csv  (in working directory)

suppressPackageStartupMessages(library(argparse))
suppressPackageStartupMessages(library(CIMLR))

# ---------- Argument parsing ----------
parser <- ArgumentParser(description = "CIMLR multi-omics clustering")
parser$add_argument("--dataset",   required = TRUE,  help = "Dataset name (e.g. ACC)")
parser$add_argument("--version",   required = TRUE,  help = "Data version: Top, Original, Aligned")
parser$add_argument("--data_path", required = TRUE,  help = "Root path to Clustering_datasets/")
parser$add_argument("--k",         type = "integer", default = 2, help = "Number of clusters (default: 2)")
args <- parser$parse_args()

dataset   <- args$dataset
version   <- args$version
data_path <- args$data_path
k         <- args$k

# ---------- File suffix per version ----------
suffix_map <- list(Original = "", Top = "_top", Aligned = "_aligned")
if (!version %in% names(suffix_map)) {
  stop(sprintf("Unknown version '%s'. Must be one of: %s", version, paste(names(suffix_map), collapse = ", ")))
}
suffix <- suffix_map[[version]]

# ---------- Build file paths ----------
base_dir <- file.path(data_path, dataset, version)

make_path <- function(omics) {
  file.path(base_dir, paste0(dataset, "_", omics, suffix, ".csv"))
}

omics_names <- c("mRNA", "miRNA", "CNV", "Methy")
paths <- setNames(lapply(omics_names, make_path), omics_names)

for (nm in omics_names) {
  if (!file.exists(paths[[nm]])) {
    stop(sprintf("File not found: %s", paths[[nm]]))
  }
}

# ---------- Read and transpose data ----------
# Input CSVs: features x samples  →  CIMLR expects samples x features
cat(sprintf("Loading %s %s data...\n", dataset, version))
mats <- lapply(paths, function(p) {
  df <- read.csv(p, row.names = 1, check.names = FALSE)
  t(as.matrix(df))   # samples x features
})

sample_names <- rownames(mats[[1]])
cat(sprintf("  Samples: %d\n", length(sample_names)))
cat(sprintf("  Features per omics: %s\n",
    paste(sapply(mats, ncol), collapse = " / ")))

# ---------- Concatenate omics ----------
# CIMLR takes a single concatenated data matrix (samples x total_features)
X <- do.call(cbind, mats)
cat(sprintf("  Combined features: %d\n", ncol(X)))

# ---------- Run CIMLR ----------
cat(sprintf("Running CIMLR with k = %d...\n", k))
result <- CIMLR(X, k, nstart = 10)

labels <- result$y$cluster

# ---------- Save results ----------
out_file <- sprintf("results_%s_%s_CIMLR.csv", dataset, version)
result_df <- data.frame(
  sample  = sample_names,
  cluster = as.integer(labels),
  row.names = NULL
)
write.csv(result_df, file = out_file, row.names = FALSE, quote = FALSE)
cat(sprintf("Results saved to: %s\n", out_file))
cat("Cluster distribution:\n")
print(table(labels))
